File: internal/shell/unix.go
//go:build darwin || linux

package shell

import (
	"os"
	"strings"

	"github.com/gvcgo/version-manager/internal/shell/sh"
)

var _ Sheller = (*Shell)(nil)

type Shell struct {
	sh.Sheller
}

func NewShell() *Shell {
	var shell sh.Sheller
	shellPath := os.Getenv("SHELL")
	switch {
	case strings.HasSuffix(shellPath, sh.Bash):
		shell = sh.NewBashShell()
	case strings.HasSuffix(shellPath, sh.Zsh):
		shell = sh.NewZshShell()
	case strings.HasSuffix(shellPath, sh.Fish):
		shell = sh.NewFishShell()
	default:
		// use bash as default shell, in case os.Getenv("SHELL") does not work in docker containers.
		shell = sh.NewBashShell()
	}
	return &Shell{shell}
}

func (s *Shell) SetPath(path string) {
	content, _ := os.ReadFile(s.VMEnvConfPath())
	data := strings.TrimSpace(string(content))

	path = s.PackPath(path)
	if !strings.Contains(data, path) {
		data = data + "\n" + path
		_ = os.WriteFile(s.VMEnvConfPath(), []byte(data), sh.ModePerm)
	}
}

func (s *Shell) UnsetPath(path string) {
	content, _ := os.ReadFile(s.VMEnvConfPath())
	data := strings.TrimSpace(string(content))

	path = s.PackPath(path)
	if strings.Contains(data, path) {
		data = strings.ReplaceAll(data, path, "")
		data = strings.ReplaceAll(data, "\n\n", "\n")
		_ = os.WriteFile(s.VMEnvConfPath(), []byte(data), sh.ModePerm)
	}
}

func (s *Shell) SetEnv(key, value string) {
	content, _ := os.ReadFile(s.VMEnvConfPath())
	data := string(content)

	env := s.PackEnv(key, value)
	if !strings.Contains(data, env) {
		data = data + "\n" + env
		_ = os.WriteFile(s.VMEnvConfPath(), []byte(data), sh.ModePerm)
	}
}

func (s *Shell) UnsetEnv(key string) {
	content, _ := os.ReadFile(s.VMEnvConfPath())
	data := string(content)
	env := s.PackEnv(key, "")
	for _, line := range strings.Split(data, "\n") {
		if strings.HasPrefix(line, env) {
			data = strings.ReplaceAll(data, line, "")
			data = strings.ReplaceAll(data, "\n\n", "\n")
		}
	}
	_ = os.WriteFile(s.VMEnvConfPath(), []byte(data), sh.ModePerm)
}

func (s *Shell) Close() {}

File: rust-vmr/crates/vmr-shell/src/unix.rs
use std::env;
use std::fs;

use crate::common::{self, BashShell, FishShell, Sheller, ZshShell};

/// Shell wraps a concrete shell implementation detected from the `SHELL`
/// environment variable.
pub struct Shell {
    inner: Box<dyn Sheller>,
}

/// Detect the current user's shell from the `SHELL` env var and return a
/// matching `Shell`.  Falls back to `BashShell` when detection fails.
pub fn new_shell() -> Shell {
    let shell_path = env::var("SHELL").unwrap_or_default();
    let inner: Box<dyn Sheller> = if shell_path.ends_with(common::BASH) {
        Box::new(BashShell)
    } else if shell_path.ends_with(common::ZSH) {
        Box::new(ZshShell)
    } else if shell_path.ends_with(common::FISH) {
        Box::new(FishShell)
    } else {
        Box::new(BashShell)
    };
    Shell { inner }
}

impl Shell {
    /// Append the packed `path` to the VM env file (unless already present).
    pub fn set_path(&self, path: &str) {
        let conf = self.inner.vm_env_conf_path();
        let data = fs::read_to_string(&conf).unwrap_or_default().trim().to_string();

        let packed = self.inner.pack_path(path);
        if !data.contains(&packed) {
            let new_data = if data.is_empty() {
                packed
            } else {
                format!("{}\n{}", data, &packed)
            };
            let _ = fs::write(&conf, new_data);
        }
    }

    /// Remove the packed `path` from the VM env file.
    pub fn unset_path(&self, path: &str) {
        let conf = self.inner.vm_env_conf_path();
        let data = fs::read_to_string(&conf).unwrap_or_default().trim().to_string();

        let packed = self.inner.pack_path(path);
        if data.contains(&packed) {
            let new_data = data
                .replace(&packed, "")
                .replace("\n\n", "\n")
                .trim()
                .to_string();
            let _ = fs::write(&conf, new_data);
        }
    }

    /// Append `key=value` (packed by the current shell) to the VM env file
    /// unless it already exists.
    pub fn set_env(&self, key: &str, value: &str) {
        let conf = self.inner.vm_env_conf_path();
        let data = fs::read_to_string(&conf).unwrap_or_default();

        let packed = self.inner.pack_env(key, value);
        if !data.contains(&packed) {
            let new_data = format!("{}\n{}", data.trim_end(), &packed);
            let _ = fs::write(&conf, new_data);
        }
    }

    /// Remove every line that starts with the packed `key`-only prefix from
    /// the VM env file.
    pub fn unset_env(&self, key: &str) {
        let conf = self.inner.vm_env_conf_path();
        let data = fs::read_to_string(&conf).unwrap_or_default();

        let prefix = self.inner.pack_env(key, "");
        let mut lines: Vec<&str> = data.lines().collect();
        lines.retain(|line| !line.starts_with(&prefix));
        let new_data = lines.join("\n").replace("\n\n", "\n").trim().to_string();
        let _ = fs::write(&conf, new_data);
    }
}
